### DISCUSSION TITLE
Confused about USERNAME_RE

### DISCUSSION THREAD

**Discussion by dullbananas** [mastodon/mastodon#26463] [Q&A] [answered]
https://github.com/mastodon/mastodon/blob/34f3001278606f6c1371e3a7f4f1f7a6c235ec9b/app/models/account.rb#L65

What is the second `[a-z0-9_]+` for? I don't see how it's possible for characters to match this instead of `[a-z0-9_.-]+`.

**Comment by trwnh** [comment]
i believe it's to allow dots and dashes only in the middle of usernames, and not at the end or beginning?

**Comment by ryanmio** [comment, marked as answer]
The second [a-z0-9_]+ is so the username always ends with an alphanumeric character or an underscore. The [a-z0-9_**.-**]+ part can include periods or dashes, which will break parsing the username. The regex makes sure the username starts and ends with [a-z0-9_] but can have [a-z0-9_.-] in the middle.

**Reply by ryanmio** [reply]
@dullbananas did this answer your question?

**Reply by dullbananas** [reply]
Yes
